common/utilities/timing.py

```diff
@@ -0,0 +1,127 @@
+import logging
+import time
+from functools import wraps
+
+
+class StopWatch:
+
+    start_time = None
+
+    def start_timer(self):
+        self.start_time = time.time()
+
+    def stop_timer(self):
+        end_time = time.time()
+        diff = end_time - self.start_time
+        return diff
+
+
+def time_method(method):
+    """
+    A method decorator that logs the time taken to execute a given method
+
+    :param method:
+    :return:
+    """
+
+    @wraps(method)
+    def invoke_method_with_timer(*args, **kwargs):
+        stopwatch = StopWatch()
+        stopwatch.start_timer()
+        try:
+            return method(*args, **kwargs)
+        finally:
+            duration = stopwatch.stop_timer()
+            duration = round(duration, 3)
+            logging.info({
+                'method': method.__name__,
+                'duration': duration
+            })
+
+    return invoke_method_with_timer
+
+
+def time_method_async(method):
+    """
+    A method decorator that logs the time taken to execute a given method. This decorator should be used to wrap
+    asynchronous methods as the standard time_method decorator does not handle async methods effectively
+    """
+
+    @wraps(method)
+    async def invoke_method_with_timer(*args, **kwargs):
+        stopwatch = StopWatch()
+        stopwatch.start_timer()
+        try:
+            return await method(*args, **kwargs)
+        finally:
+            duration = stopwatch.stop_timer()
+            duration = round(duration, 3)
+            logging.info({
+                'method': method.__name__,
+                'duration': duration
+            })
+
+    return invoke_method_with_timer
+
+
+def log_request(method):
```

Comment:
Do we need these Tornado-specific methods? From a quick check of the existing MHS code, Tornado already logs the following for me when a request completes:

```
404 POST /foo (::1) 1.29ms
```

Is the goal that we can use these to ensure a consistent logging format for request handler methods?

Reply:
As discussed these were included as something similar to what spine has. I think if tornado is logging things we probably want to ensure it also complies with the format we're wanting to log with?

Reply:
Yeah, I think as we discussed offline it is worth having these methods to allow us to log details of HTTP requests in the same format as the rest of our messages.